### PR TITLE
copy javadoc for Activity @Extra and @FragmentArg to thier Builder

### PR DESCRIPTION
**Pull request by dodgex**
see https://github.com/excilys/androidannotations/issues/1525

### CODE REVIEW

File: AndroidAnnotations/androidannotations/src/main/java/org/androidannotations/handler/FragmentArgHandler.java [hunk 1]
```diff
@@ -113,5 +113,12 @@ private void createBuilderInjectionMethod(Element element, EFragmentHolder holde
 		JVar arg = method.param(paramClass, fieldName);
 		method.body().invoke(builderArgsField, bundleHelper.getMethodNameToSave()).arg(argKeyStaticField).arg(arg);
 		method.body()._return(_this());
+
+		String docComment = processingEnv.getElementUtils().getDocComment(element);
+		if (docComment != null) {
+			method.javadoc().append(docComment.replaceAll("\r", "").trim());
```
Comment (WonderCsabo) [inline AndroidAnnotations/androidannotations/src/main/java/org/androidannotations/handler/FragmentArgHandler.java]: Can you explain this line?

Comment (dodgex) [inline AndroidAnnotations/androidannotations/src/main/java/org/androidannotations/handler/FragmentArgHandler.java]: i assume you ask for the `.replaceAll("\r", "")`. I removed the `\n` because it breaks the output.

e.g.
```
/**
 * Hello
World
 *
 */
```
instead of
```
/**
 * Hello
 * World
 *
 */
```

Comment (WonderCsabo) [inline AndroidAnnotations/androidannotations/src/main/java/org/androidannotations/handler/FragmentArgHandler.java]: You mean you removed the `\r`, right? But why is this needed? CodeModel assumes only `\n` line ending and break with `\r\n`?

Comment (dodgex) [inline AndroidAnnotations/androidannotations/src/main/java/org/androidannotations/handler/FragmentArgHandler.java]: Yes i mean that i removed the `\r`, sorry.

the `.getDocComment(element)` call returns comment with `\r\n` but when i apped that to the javadoc i get the line-break without an `*` in the line (like the first snippet). removing the `\r` resulted in having a correctly formatted javadoc output.

Comment (WonderCsabo) [inline AndroidAnnotations/androidannotations/src/main/java/org/androidannotations/handler/FragmentArgHandler.java]: I see. Can you move this duplicated code then to `APTCodeModelHelper`?

Comment (dodgex) [inline AndroidAnnotations/androidannotations/src/main/java/org/androidannotations/handler/FragmentArgHandler.java]: updated

Comment (WonderCsabo) [inline AndroidAnnotations/androidannotations/src/main/java/org/androidannotations/handler/FragmentArgHandler.java]: I would extract this whole block into one method:

if (docComment != null) {   method.javadoc().append(codeModelHelper.
trimDocComment(docComment));  }

Comment (WonderCsabo) [inline AndroidAnnotations/androidannotations/src/main/java/org/androidannotations/handler/FragmentArgHandler.java]: I would extract this whole block into one method:
```java
if (docComment != null) {
  method.javadoc().append(docComment.replaceAll("\r", "").trim());
}
```

Comment (dodgex) [inline AndroidAnnotations/androidannotations/src/main/java/org/androidannotations/handler/FragmentArgHandler.java]: done

**Comment by WonderCsabo** [comment]
Thanks! Seems to be okay. After the plugin system is merged, i will have to ask you to rebase this unfortunately.

**Comment by dodgex** [comment]
I'm just waiting for the moment i can rebase ;-)

**Comment by WonderCsabo** [comment]
Please rebase this onto `develop`.

**Comment by dodgex** [comment]
rebased

**Comment by WonderCsabo** [comment]
It seems you have compilation errors.

**Comment by dodgex** [comment]
yeah... currently fighting with my eclipse to get AA working.

**Comment by WonderCsabo** [comment]
Sorry... We changed to IntelliJ. I did not update the instructions, yet.

**Comment by dodgex** [comment]
the androidannotations-test project is missing project.properties :/

**Comment by dodgex** [comment]
should be fixed

**Comment by dodgex** [comment]
TODO: copy ServiceAction javadoc

**Comment by WonderCsabo** [comment]
@dodgex will you add `@ServiceAction` javadoc gen in this PR?

**Comment by dodgex** [comment]
yeah. it is planned. but i had no time yet. maybe this evening.

**Comment by WonderCsabo** [comment]
OK, i will wait with merging, then.

**Comment by dodgex** [comment]
PR updated with javadoc for `@ServiceAction`. here it is a plain copy of the original doc.

**Comment by WonderCsabo** [comment]
Thanks for handling `@ServiceAction` as well. And it seems it also handles javadoc for parameters correctly:

```java
/**
 * this is a javadoc comment
 *
 * @param param This is something
 */
@ServiceAction
void action(int param) {
}
```
```java
/**
 * this is a javadoc comment
 * 
 *  @param param This is something
 * 
 */
public ServiceWithServiceAction_.IntentBuilder_ action(int param) {
	action(ACTION);
	super.extra(PARAM_EXTRA, param);
	return this;
}
```

Can i ask you to add the param to the test? Also please add a `@return` tag to the `ServiceAction` javadoc as you did for the other two annotations (and also test that)..

Thanks for your work!

**Comment by WonderCsabo** [comment]
Well, the only little problem with `@param` that it is has extra space before it. But i think that is okay, since JavaDoc parsers should be able to handle that (IntelliJ's can).

**Comment by dodgex** [comment]
no idea where the space comes from. but i'm pretty sure that JavaDoc parsers do not care for spaces at all (if not in a `<pre>` block)

**Comment by dodgex** [comment]
PR updated

**Comment by WonderCsabo** [comment]
I know where it comes from. The javadoc which is added [here](https://github.com/dodgex/androidannotations/commit/425c958ed43431b42c73c3b676c97f25b508603b#diff-8978079fb392ef4e876921a75c72d549R126), is interpreted as the main JavaDoc. When codemodel renders it, it adds a space after the starts. But the `@param` already has a space, that is why it is duplicated. But this is not a problem.

**Comment by WonderCsabo** [comment]
After the final review of this PR, i realized there is another last annotations, which can use JavaDoc copy: `@SharedPref` and `@DefaultXXX`. These are not so important, but if you want to contribute that too, go ahead.

**Comment by WonderCsabo** [comment]
Follow up issue for shared prefs: https://github.com/excilys/androidannotations/issues/1551.
